cas/tree.go

```diff
@@ -0,0 +1,323 @@
+package cas
+
+import (
+	"archive/tar"
+	"crypto/sha512"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+
+	specaci "github.com/coreos/rocket/Godeps/_workspace/src/github.com/appc/spec/aci"
+	"github.com/coreos/rocket/Godeps/_workspace/src/github.com/appc/spec/pkg/tarheader"
+	"github.com/coreos/rocket/Godeps/_workspace/src/github.com/appc/spec/schema/types"
+	"github.com/coreos/rocket/pkg/aci"
+)
+
+const (
+	hashfilename     = "hash"
+	renderedfilename = "rendered"
+)
+
+// A tree store represent a store of rendered ACIs
+// The image's key becomes the name of the directory containing the rendered aci.
+type TreeStore struct {
+	path string
+}
+
+// Write renders the ACI with the provided key in the treestore
+// Write, to avoid having a rendered ACI with old stale files, requires that
+// the destination directory doesn't exists (usually Remove should be called
+// before Write)
+func (ts *TreeStore) Write(key string, ds *Store) error {
+	treepath := filepath.Join(ts.path, key)
+	fi, _ := os.Stat(treepath)
+	if fi != nil {
+		return fmt.Errorf("treestore: path %s already exists", treepath)
+	}
+	imageID, err := types.NewHash(key)
+	if err != nil {
+		return fmt.Errorf("treestore: cannot convert key to imageID: %v", err)
+	}
+	err = aci.RenderACIWithImageID(*imageID, treepath, ds)
+	if err != nil {
+		return fmt.Errorf("treestore: cannot render aci: %v", err)
+	}
+	hash, err := ts.Hash(key)
+	if err != nil {
+		return fmt.Errorf("treestore: cannot calculate tree hash: %v", err)
+	}
+	err = ioutil.WriteFile(filepath.Join(treepath, hashfilename), []byte(hash), 0644)
+	if err != nil {
+		return fmt.Errorf("treestore: cannot write hash file: %v", err)
+	}
+	// before creating the "rendered" flag file we need to ensure that all data is fsynced
+	filepath.Walk(treepath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		f, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		err = f.Sync()
+		if err != nil {
+			return err
+		}
+		return nil
+	})
```

Comment:
@iaguis updated adding tests for a symlink, a dangling symlinks and a named pipe. Also fixed the buildWalker to sync with appc/spec/aci (the copy is due to the need to ignore also hashfile and renderedfile, need to find a way to avoid it).